File: rl_py/agent.py
#!/usr/bin/env python

import random

import rospy

from rl_py.msg import (RLEnvDescription, RLStateReward, RLEnvSeedExperience,
                       RLAction, RLExperimentInfo)
from rl_agent.cli_args import parser
from rl_agent.Sarsa import Sarsa
from rl_agent.Agent import Agent

NODE = "RLAgent"

out_rl_action = None
out_exp_info = None

first_action = True

agent = None
args = None

info = RLExperimentInfo()
rng = None
# int model = C45TREE;
# int explore = GREEDY;
# int modelcombo = BEST;
# int planner = PAR_ETUCT_ACTUAL;


# Process the state/reward message from the environment
def process_state(state_in):
    global info, first_action

    if agent == None:
        print("No agent yet")
        return

    a = RLAction()

    # First action
    if first_action:
        a.action = agent.first_action(state_in.state);
        info.episode_reward = 0;
        info.number_actions = 1;
    else:
        info.episode_reward += state_in.reward;
        # if terminal, no action, but calculate reward sum
        if state_in.terminal:
            agent.last_action(state_in.reward);
            print("Episode", info.episode_number, "reward:", info.episode_reward)
            # publish episode reward message
            out_exp_info.publish(info);

            info.episode_number += 1
            info.episode_reward = 0;
            first_action = True;
            return
        else:
            a.action = agent.next_action(state_in.reward, state_in.state)
            info.number_actions += 1

    first_action = false;

    # publish agent's action
    out_rl_action.publish(a);


# Process seeds for initializing model
def processSeed(seed_in):
    if agent == None:
        print("No agent yet")
        return

    seeds = []
    seed1 = Experience()
    seed1.s = seed_in.from_state;
    seed1.next = seed_in.to_state;
    seed1.act = seed_in.action;
    seed1.reward = seed_in.reward;
    seed1.terminal = seed_in.terminal;
    seeds.append(seed1);

    agent.seed_exp(seeds);

# Process the env description message from the environment
def process_env_description(env_in):
    global agent, first_action, info
    # Initialize the agent based on some info from the environment descriptor
    agent = None;

    if args.type == "qlearner":
        print("Agent: QLearner")
        agent = QLearner(num_actions=env_in.num_actions,
                         gamma=args.gamma,
                         initialvalue=args.initialvalue,
                         alpha=args.alpha,
                         epsilon=args.epsilon,
                         rng=rng)
    elif ards.type == "sarsa":
        print("Agent: Sarsa")
        agent = Sarsa(num_actions=env_in.num_actions,
                      gamma=args.gamma,
                      initialvalue=args.initialvalue,
                      alpha=args.alpha,
                      epsilon=args.epsilon,
                      _lambda=args._lambda,
                      rng=rng)
    else:
        print("Invalid Agent!")
        sys.exit(-1)

  #   Agent* a2 = agent;
  # // not for model based when doing continuous model
  # if (nstates > 0 && (model != M5ALLMULTI || strcmp(agentType, "qlearner") == 0)){
  #   int totalStates = powf(nstates,envIn->min_state_range.size());
  #   if (PRINTS) cout << "Discretize with " << nstates << ", total: " << totalStates << endl;
  #   agent = new DiscretizationAgent(nstates, a2,
  #                                   envIn->min_state_range,
  #                                   envIn->max_state_range, PRINTS);
  # }

    first_action = True;
    info.episode_number = 0;
    info.episode_reward = 0;

if __name__ == "__main__":
    args = parser.parse_args()
    rng = random.Random()
    queue_depth = 1

    # Change to dictionarized exploration
    if args.exploration == "unknown":
        args.exploration = "EXPLORE_UNKNOWN"
    elif args.exploration == "greedy":
        args.exploration = "GREEDY"
    elif args.exploration == "epsilongreedy":
        args.exploration = "EPSILONGREEDY"
    elif args.exploration == "unvisitedstates":
        args.exploration = "UNVISITED_BONUS"
    elif args.exploration == "unvisitedactions":
        args.exploration = "UNVISITED_ACT_BONUS"
    elif args.exploration == "variancenovelty":
        args.exploration = "DIFF_AND_NOVEL_BONUS"

    if args.type != "sarsa":
        raise NotImplementedError()

    # Make a node
    rospy.loginfo("Making a node ...")
    rospy.init_node(NODE, anonymous=True)

    # Publishers
    rospy.loginfo("Making publishers ...")
    out_rl_action = rospy.Publisher("rl_agent/rl_action", RLAction, queue_size=queue_depth, latch=False)
    out_exp_info = rospy.Publisher("rl_agent/rl_experiment_info", RLExperimentInfo, queue_size=queue_depth, latch=False)

    # Subscribers
    rospy.loginfo("Making subscribers ...")
    rl_description = rospy.Subscriber("rl_env/rl_env_description", queue_depth, process_env_description, tcp_nodelay=True)
    rl_state = rospy.Subscriber("rl_env/rl_state_reward", queue_depth, process_state, tcp_nodelay=True)
    rl_seed = rospy.Subscriber("rl_env/rl_seed", queue_depth, process_seed, tcp_nodelay=True)

    # Setup RL World
    rospy.loginfo("Setting up RL Agent ...")
    rng.seed(1+args.seed)
    init_environment(rng)

    rospy.loginfo(NODE + ": starting main loop")
    rospy.spin()

File: rl_agent/src/agent.cpp
#include <getopt.h>
#include <ros/ros.h>

#include <rl_common/core.hh>

// Messages
#include <rl_common/RLStateReward.h>
#include <rl_common/RLAction.h>
#include <rl_common/RLExperimentInfo.h>

// Agents
#include <rl_agent/Pegasus.hh>


static ros::Publisher out_rl_action;
static ros::Publisher out_exp_info;


const int MAX_STEPS = 10000;
Agent* agent = NULL;
int seed = 1;

rl_common::RLExperimentInfo info;
std::string agent_type = "";

void display_help(){
  std::cout << "\n agent --agent type [options]\n";
  std::cout << "\n Options:\n";
  std::cout << "--agent type (Agent types: pegasus)\n";
  exit(-1);
}

void init_agent() {
  agent = NULL;

  if (agent_type == "pegasus"){
    std::cout << "Agent: Pegasus" << std::endl;
    // For now, we arent using these args. Theyre reset in the constructor
    agent = new Pegasus(4, // Num state
                        2 // Num action
                        );
  } else {
    std::cout << "Invalid Agent!" << std::endl;
    display_help();
  }

  info.number_actions = 0;
}

void process_state(const rl_common::RLStateReward::ConstPtr &state_in){
  /** Process the state/reward message from the environment */

  if (agent == NULL) {
    init_agent();
  }

  rl_common::RLAction msg;

  if (info.number_actions == 0) {
    msg.action = agent->first_action(state_in->state);
    info.episode_reward = 0;
    info.number_actions += 1;

  } else if (state_in->terminal || info.number_actions > MAX_STEPS) {
    info.episode_reward += state_in->reward;
    info.episode_number += 1;
    agent->last_action(state_in->reward);
    out_exp_info.publish(info); // Publish end of episode message

    // std::cout << "RL AGENT: Episode " << info.episode_number
    //           << ", #Actions " << info.number_actions
    //           << ", Episode Reward: " << info.episode_reward << "\n";

    info.number_actions = 0;
    info.episode_reward = 0;
    return;

  } else {
    info.episode_reward += state_in->reward;
    info.number_actions += 1;
    msg.action = agent->next_action(state_in->reward, state_in->state);
  }

  out_rl_action.publish(msg);
}

int main(int argc, char *argv[]) {
  ros::init(argc, argv, "RLAgent");
  ros::NodeHandle node;

  char ch;
  const char* optflags = "as";
  int option_index = 0;
  static struct option long_options[] = {
    {"seed", 1, 0, 's'},
    {"agent", 1, 0, 'a'},
    {NULL, 0, 0, 0}
  };

  while(-1 != (ch = getopt_long_only(argc, argv, optflags, long_options, &option_index))) {
    switch(ch) {

    case 's':
      seed = std::atoi(optarg);
      std::cout << "Using seed: " << seed << "\n";
      break;

    case 'a':
      agent_type = optarg;
      std::cout << "Using agent: " << agent_type << "\n";
      break;

    default:
      display_help();
      break;
    }
  }

  if (agent_type == "") {
    display_help();
  }

  std::cout << "RL AGENT:  Initializing ROS ...\n";
  // Publishers
  out_rl_action = node.advertise<rl_common::RLAction>("rl_agent/rl_action",
                                                    1,
                                                    false);
  out_exp_info = node.advertise<rl_common::RLExperimentInfo>(
    "rl_agent/rl_experiment_info",
    1,
    false);

  // Subscribers
  ros::TransportHints no_delay = ros::TransportHints().tcpNoDelay(true);
  ros::Subscriber rl_state =  node.subscribe("rl_env/rl_state_reward",
                                             1,
                                             process_state,
                                             no_delay);

  ROS_INFO("RL AGENT: starting main loop");
  ros::spin(); // handle incoming data

  return 0;
}
